client/modules/PaymentDetails/PaymentDetails.js

```diff
@@ -0,0 +1,333 @@
+
+const nonDigit = /[^0-9]/g;
+
+class PaymentDetails extends Component
+{
+  constructor(props)
+  {
+    super(props);
+    this.state = {
+      // credit card fields
+      cardNumber: '',
+      nameOnCard: '',
+      ccv: '',
+      expiryMonth: '',
+      expiryYear: '',
+      // page state
+      updated: false,
+      expMonthDropdownOpen: false,
+      expYearDropdownOpen: false,
+      touched: {
+        cardNumber: false,
+        nameOnCard: false,
+        ccv: false,
+        expiry: false,
+      },
+    };
+  }
+
+  valids = {};
+
+  validate() {
+    const cardNumber = this.state.cardNumber.replace(nonDigit, '');
+    const cardType = ccCheck.determineCardType(cardNumber);
+    const ccv = this.state.ccv.replace(nonDigit, '');
+
+    const valids = {
+      cardNumber: cardType && ccCheck.luhn(cardNumber),
+      nameOnCard: this.state.nameOnCard.length > 0,
+      ccv: ccCheck.doesCvvMatchType(ccv, cardType),
+      expiry: !ccCheck.isExpired(this.state.expiryMonth, this.state.expiryYear),
+    }
+    console.log(valids);
+    return valids;
+  }
+
+  handleInputChange(evt) {
+    let field = evt.target.id;
+    let value = evt.target.value
+    let props = {};
+    props [field] = value;
+    props.touched = this.state.touched;
+    props.touched [field] = true;
+    this.setState(props);
+    console.log(this.state.touched);
+  }
+
+  handleExpiryMonthChange(evt) {
+    let props = {
+      expiryMonth: evt.target.innerText,
+      expMonthDropdownOpen: false }
+    props.touched = this.state.touched;
+    props.touched.expiry = true;
+    this.setState(props);
+    console.log(this.state.touched);
+  }
+
+  handleExpiryYearChange = (evt) => {
+    let props = {
+      expiryYear: evt.target.innerText,
+      expYearDropdownOpen: false }
+    props.touched = this.state.touched;
+    props.touched.expiry = true;
+    this.setState(props);
+    console.log(this.state.touched);
+  }
+
+  toggleExpMonthDropdown() {
+    this.setState({
+      expMonthDropdownOpen: !this.state.expMonthDropdownOpen,
+    });
+  }
+
+  toggleExpYearDropdown() {
+    this.setState({
+      expYearDropdownOpen: !this.state.expYearDropdownOpen,
+    });
+  }
+
+  handleSubmit(evt) {
+    evt.preventDefault();
+
+    const cardNumber = this.state.cardNumber.replace(nonDigit, '');
+    const ccv = this.state.ccv.replace(nonDigit, '');
+
+    if (this.formIsValid()) {
+      http.client().put('/paymentDetals/change', {
+        cardNumber: cardNumber,
+        nameOnCard: this.state.nameOnCard,
+        ccv: ccv,
+        expiryMonth: Number(this.state.expiryMonth),
+        expiryYear: Number(this.state.expiryYear),
+      })
+      .then(res => console.log(res))
+      .catch(err => console.log(err));
+    }
+  }
+
+  isValid(key) {
+    const valid = this.valids[key];
+    const touched = this.state.touched[key];
+    return valid || !touched;
+  }
+
+  formIsValid() {
+    return Object.keys(this.valids).some(field => this.valids[field] === false);
```

Comment:
![image](https://user-images.githubusercontent.com/37112057/38500309-9c1cda00-3c3c-11e8-9fd4-e61ace69971c.png)
i tried a quick example of this validation in the dev tools console. its not returning the value you would expect. array.some() checks for existence of at least 1 element that matched the predicate, so if one field in valids == false, then it returns true

can we change it to Object.keys(this.valids).every(field => this.valids[field] === true)

Reply:
or even change it to **!**Object.keys(this.valids).some(field => this.valids[field] === false);